tico/experimental/quantization/ptq/wrappers/fairseq/quant_encoder.py

```diff
@@ -0,0 +1,333 @@
+
+        assert hasattr(fp_encoder, "padding_idx")
+        self.padding_idx: int = int(fp_encoder.padding_idx)  # type: ignore[arg-type]
+
+        # scale = sqrt(embed_dim) unless disabled
+        embed_dim = int(self.embed_tokens.embedding_dim)  # type: ignore[arg-type]
+        no_scale = bool(getattr(self.cfg, "no_scale_embedding", False))
+        self.embed_scale: float = 1.0 if no_scale else math.sqrt(embed_dim)
+
+        # Positional embeddings (keep as-is; no FQ)
+        self.embed_positions = getattr(fp_encoder, "embed_positions", None)
+        # Optional embedding LayerNorm
+        self.layernorm_embedding = getattr(fp_encoder, "layernorm_embedding", None)
+
+        # Final encoder LayerNorm (pre-norm stacks may set this to None)
+        self.layer_norm = getattr(fp_encoder, "layer_norm", None)
+
+        # Max positions (reuse for API parity)
+        self.max_source_positions: int = int(fp_encoder.max_source_positions)  # type: ignore[arg-type]
+
+        # --- wrap encoder layers with PTQWrapper ----------------------------
+        assert hasattr(fp_encoder, "layers")
+        fp_layers = list(fp_encoder.layers)  # type: ignore[arg-type]
+        self.layers = nn.ModuleList()
+
+        # Prepare child QuantConfig namespaces: layers/<idx>
+        layers_qcfg = qcfg.child("layers") if qcfg else None
+        for i, layer in enumerate(fp_layers):
+            child_cfg = layers_qcfg.child(str(i)) if layers_qcfg else None
+            self.layers.append(
+                PTQWrapper(layer, qcfg=child_cfg, fp_name=f"{fp_name}.layers.{i}")
+            )
+
+        # Version buffer (keep for state_dict parity)
+        version = getattr(fp_encoder, "version", None)
+        if isinstance(version, torch.Tensor):
+            self.register_buffer("version", version.clone(), persistent=False)
+        else:
+            self.register_buffer("version", torch.tensor([3.0]), persistent=False)
+
+    # ----------------------------------------------------------------------
+    def forward_embedding(
+        self, src_tokens: Tensor, token_embedding: Optional[Tensor] = None
+    ) -> Tuple[Tensor, Tensor]:
+        """
+        Embed tokens and add positional embeddings. Dropout/quant_noise are removed.
+        Returns:
+            x (B, T, C), embed (B, T, C)  # embed is the token-only embedding
+        """
+        if token_embedding is None:
+            token_embedding = self.embed_tokens(src_tokens)
+        embed = token_embedding  # token-only
+
+        x = self.embed_scale * token_embedding
+        if self.embed_positions is not None:
+            x = x + self.embed_positions(src_tokens)
+        if self.layernorm_embedding is not None:
+            x = self.layernorm_embedding(x)
+        # No dropout, no quant_noise here (inference-only)
+        return x, embed
+
+    # ----------------------------------------------------------------------
+    def forward(
```

Comment:
The original implementation is in [here](https://github.com/facebookresearch/fairseq/blob/d13e14a800bb588e5a77fb4e551f554ff9b24a72/fairseq/models/transformer/transformer_encoder.py#L173).